db/bbolt.go

```diff
@@ -64,80 +64,68 @@ func (bb *BBoltHandler) CreateInitialSchema() error {
 	})
 }
 
-func (bb *BBoltHandler) InsertBlock(block *types.Block) error {
-	bb.logger.Info("Inserting block to DB", zap.Uint64("block_height", block.BlockHeight))
-
-	// Store mapping number -> block
-	err := bb.db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket([]byte(blocksBucket))
-		key := bb.itob(block.BlockHeight)
-		blockBytes, err := json.Marshal(block)
-		if err != nil {
-			return err
-		}
-		return b.Put(key, blockBytes)
-	})
-	if err != nil {
-		bb.logger.Error("Error inserting block", zap.Error(err))
-		return err
+func (bb *BBoltHandler) InsertBlocks(blocks []*types.Block) error {
+	if len(blocks) == 0 {
+		return nil
 	}
 
-	// Store mapping hash -> number
-	err = bb.db.Update(func(tx *bolt.Tx) error {
-		b := tx.Bucket([]byte(blockHeightsBucket))
-		return b.Put([]byte(block.BlockHash), bb.itob(block.BlockHeight))
-	})
-	if err != nil {
-		bb.logger.Error("Error inserting block", zap.Error(err))
-		return err
-	}
+	bb.logger.Info("Batch inserting blocks to DB", zap.Int("count", len(blocks)))
+
+	// Single transaction for all operations
+	return bb.db.Update(func(tx *bolt.Tx) error {
+		blocksBucket := tx.Bucket([]byte(blocksBucket))
+		heightsBucket := tx.Bucket([]byte(blockHeightsBucket))
+		indexBucket := tx.Bucket([]byte(indexerBucket))
+
+		var minHeight, maxHeight uint64 = math.MaxUint64, 0
+
+		// Insert all blocks
+		for _, block := range blocks {
+			// Update min/max heights
+			if block.BlockHeight < minHeight {
+				minHeight = block.BlockHeight
+			}
+			if block.BlockHeight > maxHeight {
+				maxHeight = block.BlockHeight
+			}
+
+			// Store block data
+			blockBytes, err := json.Marshal(block)
+			if err != nil {
+				return err
+			}
+			if err := blocksBucket.Put(bb.itob(block.BlockHeight), blockBytes); err != nil {
+				return err
+			}
 
-	// Get current earliest block
-	// If it is unset, update earliest block
-	earliestBlock, err := bb.QueryEarliestFinalizedBlock()
-	if earliestBlock == nil || errors.Is(err, types.ErrBlockNotFound) {
-		err = bb.db.Update(func(tx *bolt.Tx) error {
-			b := tx.Bucket([]byte(indexerBucket))
-			return b.Put([]byte(earliestBlockKey), bb.itob(block.BlockHeight))
-		})
-		if err != nil {
-			bb.logger.Error("Error updating earliest block", zap.Error(err))
-			return err
+			// Store height mapping
+			if err := heightsBucket.Put([]byte(block.BlockHash), bb.itob(block.BlockHeight)); err != nil {
+				return err
+			}
 		}
-	}
-	if err != nil {
-		bb.logger.Error("Error getting earliest block", zap.Error(err))
-		return err
-	}
 
-	// Get current latest block
-	latestBlock, err := bb.QueryLatestFinalizedBlock()
-	if latestBlock == nil {
-		latestBlock = &types.Block{BlockHeight: 0}
-	}
-	if err != nil {
-		bb.logger.Error("Error getting latest block", zap.Error(err))
-		return err
-	}
+		// Update earliest block if needed
+		earliestBytes := indexBucket.Get([]byte(earliestBlockKey))
```

Comment:
the old code calls `QueryEarliestFinalizedBlock`. is it b/c we don't need to get the actual block so we don't need to call it here?

Reply:
yeah its just a more direct way to query. `QueryEarliestFinalizedBlock` is meant to be an external api, so here we can just get the marshalled block directly and check if it exists.